Add more momentum diagnostics

owang01 opened the pull request:
## What changes does this PR introduce?
New feature to output more momentum diagnostics


## What is the current behaviour? 
(You can also link to an open issue here)


## What is the new behaviour 
Added more momentum diagnostics, originally implemented by Helen Pillar @hpillar (and probably @antnguyen13) for checkpoint68d, to further decompose momentum diagnostics of advection, explicit dissipation, and external forcing to different contributions.


## Does this PR introduce a breaking change? 
No


## Other information:
Diagnostics for 3d U/V momentum tendency from bottom drag `UBotDrag` and `VBotDrag`, removed in PR #219, are put back. 

The four new diagnostics `U/Vm_hDis2` and `U/Vm_hDis4` are only active if using Vector-Inv form (i.e. `ALLOW_MOM_VECINV` is defined) and only if `useStrainTensionVisc=.FALSE.` These four new diagnostics are filled in S/R `MOM_VI_HDISSIP`, which is called only if `useStrainTensionVisc=.FALSE.` 

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

jm-c commented:
I did not look into the details, but since

1.  there are many more diagnostics added here (20 ?) and
2. some involves adding back diagnostics that have been removed (e.g, in PR #219)
3. some pieces of code are not very clear to me (3-D diag for wind ? that also include tides ?)

it mind be better to open an issue first on what diagnostics we would like to add  and  which one we don't necessary need if we know how to compute from other diags.

owang01 commented:
Thanks @jm-c for your comment. I will open an issue.

jm-c commented on pkg/mom_vecinv/mom_vecinv.F:
```diff
@@ -130,6 +137,14 @@ SUBROUTINE MOM_VECINV(
       _RL viscAh_D(1-OLx:sNx+OLx,1-OLy:sNy+OLy)
       _RL viscA4_Z(1-OLx:sNx+OLx,1-OLy:sNy+OLy)
       _RL viscA4_D(1-OLx:sNx+OLx,1-OLy:sNy+OLy)
+#if ( defined ALLOW_MOM_COMMON && defined ALLOW_MOM_COMMON )
```
A quick comment: this line could just be replaced by "#ifdef ALLOW_MOM_TEND_EXTRA_DIAGS" (and no need to check for ALLOW_MOM_COMMON since `pkg/mom_vecinv` will not be compiled without `pkg/mom_common`, see `pkg/pkg_depend` dependence file).
And I think this typo/mistake is repeated several times.

owang01 replied:
@jm-c Thank you for your comment. I have replaced `#if ( defined ALLOW_MOM_COMMON && defined ALLOW_MOM_COMMON )` with `#ifdef ALLOW_MOM_TEND_EXTRA_DIAGS`. In addition, I have removed two occurrences of "#ifdef ALLOW_MOM_COMMON" in `pkg/mom_vecinv/mom_vecinv.F`.

jm-c commented on pkg/mom_vecinv/mom_vecinv.F:
```diff
@@ -198,6 +209,15 @@ SUBROUTINE MOM_VECINV(
         zStar(i,j) = 0.
         guDiss(i,j)= 0.
         gvDiss(i,j)= 0.
+#ifdef ALLOW_MOM_TEND_EXTRA_DIAGS
```
We could think of combining these 2 CPP ifdef on a single line, since they appear muttiple times in this file.

owang01 replied:
I combined the two CPP ifdefs (`ALLOW_MOM_TEND_EXTRA_DIAGS` and `ALLOW_DIAGNOSTICS`) into one in the new commit that I'm going to check in shortly.

jm-c reviewed (commented):
I took a quick look and this looks on the right track. Just few comments:

1. Regarding the name of harmonic & biharmonic momentum diagnostics ("U/Vm_Diss2/4"): we may want to add something related to "horizontal", since the vertical (hamonic) viscosity is not part of "U/Vm_Diss2" (we have a biharmonic verticall diffusion but, currently, no biharmonic vertical viscosity). May be "U/Vm_hDis2" and "U/Vm_hDis4" ? 
2. And regarding these 4 new diags: they are only filled if using Vector-Inv form and only if useStrainTensionVisc=F. I think it would be good to mention this second limitation in PR description (+ where these diags are defined ?). And regarding the first limitation, it would require much more code to get theses working in flux-form version (both Harmonic and bi-harmonic are combine into a single flux before taking the divergence) so may be we could enclose these 4 diags within "#ifdef ALLOW_MOM_VECINV" to clarify this point ?
3. I could imagine some ways to reduce the number of new local array but will need to check if it might work. I can suggest that I will explore this a bit later.

owang01 commented on pkg/mom_common/mom_diagnostics_init.F:
```diff
@@ -382,6 +382,43 @@ SUBROUTINE MOM_DIAGNOSTICS_INIT( myThid )
      I   diagName, diagCode, diagUnits, diagTitle, diagMate, myThid )
 
+c -----------------------------------------------------------------------------------------------------
+cHP: adding new diagnostics for full momentum budget analysis
```
Done.

owang01 commented:
@jm-c Thank you for your review and comments. I have addressed all your comments in the new commit [d80a3b3](https://github.com/MITgcm/MITgcm/pull/817/commits/d80a3b3b2255bebf013d79b07ef4c962717a9f48).

owang01 commented:
@jm-c Thank you for the changes—they look good to me. I don’t mean to be picky, but I’ve noticed some inconsistency in the spacing format across several `CALL DIAGNOSTICS_FILL` statements. Here are two examples:
```
CALL DIAGNOSTICS_FILL( uDissip,'Um_hDis2', k,1,2,bi,bj, myThid )
CALL DIAGNOSTICS_FILL( uD4, 'Um_hDis4', k,1,2, bi,bj, myThid )
```

jm-c commented:
@owang01 You are right. I guess I was reluctant to split these calls over 2 lines but we could also keep "k,1,2,bi,bj" together (no spacing) in all these new DIAGNOSTICS_FILL calls.

jm-c commented:
@owang01 I pushed little (minor) adjustments to these 2 files (but did not touch the previous DIAGNOSTICS_FILL calls to facilitate merging with PR #809 which also affects `mom_vecinv.F`). 
I might switch on the new CPP option in one of the verification exp. so that this code get compiled.

jm-c commented:
From my side this PR is ready to be merged. @owang01 if you agree, and assuming no one else wants to add anything here, we could merge it soon after.

owang01 commented:
@jm-c LGTM. Thank you for the additional changes.
